examples/DivisorCurve.jl

```diff
@@ -0,0 +1,329 @@
+export AffineCurveDivisor, ProjCurveDivisor, iseffective, multiplicity,
+       divisor, curve, degree
+
+################################################################################
+
+abstract type CurveDivisor end
+
+################################################################################
+# The curve C is assume to be smooth and irreducible, but it is not checked.
+@doc Markdown.doc"""
+    AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+struct AffineCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::AffinePlaneCurve{S}
+    divisor::Dict{Point{S}, Int}
+    degree::Int
+    function AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            ison_curve(P, C) || error("The point ", P.coord, " is not on the curve")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+
+function AffineCurveDivisor(C::AffinePlaneCurve{S}, P::Point{S}, m::Int=1) where S <: FieldElem
+    return AffineCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+struct ProjCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::ProjPlaneCurve{S}
+    divisor::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}
+    degree::Int
+    function ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            ison_curve(P, C) || error("The point ", P, " is not on the curve")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+#function Base.show(io::IO, D::ProjCurveDivisor)
+#    println(io, "On the curve ", D.C.eq, ", divisor defined by ", D.divisor)
+#end
+
+function ProjCurveDivisor(C::ProjPlaneCurve{S}, P::Oscar.Geometry.ProjSpcElem{S}, m::Int=1) where S <: FieldElem
+    return ProjCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+# To get a nice output for the divisors. The needs_parentheses is needed since
+# expressify wants to know if parentheses are needed.
+
+function AbstractAlgebra.needs_parentheses(A::Array{T, 1}) where T <: RingElem
+    return false
+end
+
+function AbstractAlgebra.expressify(@nospecialize(a::AffineCurveDivisor); context = nothing)
+   prod = Expr(:call, :+)
+   for (p, i) in a.divisor
```

Comment:
Optional comment:

Perhaps use `(P,m)` instead of `(p,i)` for consistency? (I started to use `(P,m)` in my earlier suggestions based on what you used in one of the constructors).